### ISSUE TITLE
Nestes tabs in BO issue

### ISSUE THREAD

**Issue by kpodemski** [PrestaShop/PrestaShop#20908]
#### Describe the bug

We can't (or i'm doing it wrong) create nested tabs properly using new tab registration method. My menu:

![image](https://user-images.githubusercontent.com/2137763/92539313-6173da80-f241-11ea-983c-8837800a94bc.png)

In my opinion, there are two issues:

- we should be able to use something like `skip_validation`, to omit validation here:
https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/Module/Tab/ModuleTabRegister.php#L218

I don't want to have controller for parent tab.

- we should be able to omit this part of code:
https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/Module/Tab/ModuleTabRegister.php#L400

this causing an error with duplicated parent.

Possible solutions?

- add `skip_validation` param
- add `skip_duplication` param? for that one i'm not sure, it's not intuitive that we need this in second tab, not a first one, i'm open for suggestions here

Let me know what y'all think, i'll do a proper PR if needed.

#### Expected behavior

I want my menu to look like this:

![image](https://user-images.githubusercontent.com/2137763/92539287-4ef9a100-f241-11ea-8f07-9ba8fe68f6e7.png)

where "Parent tab" is linking to first child. I've successfully created this structure with suggested modifications to `ModuleTabRegister`.

#### Steps to Reproduce

1. Install example module: https://github.com/kpodemski/kpnestedtabs
2. See that tabs are "broken".
#### Additional information

* PrestaShop version: 1.7
* PHP version: N/A

**Comment by kpodemski** [comment]
Proposed syntax: 

https://github.com/kpodemski/kpnestedtabs/blob/proposed-changes/kpnestedtabs.php#L11

#notifyme

**Comment by khouloudbelguith** [comment]
Hi @kpodemski,

I tried with another module, for example, this [module](https://addons.prestashop.com/fr/blog-forum-actualites/41015-blog-pro-programme-de-fidelite.html) and it is ok
![image](https://user-images.githubusercontent.com/16067358/92575167-4d070080-f288-11ea-9f64-362c446e30ff.png)
On this github repository there is no software support. This is an open source project.


You can find help

- on PrestaShop forums https://www.prestashop.com/forums/
- or by purchasing software support plan from https://www.prestashop.com/en/support

Thanks!

**Comment by kpodemski** [comment]
@khouloudbelguith 

oh, wow :) that was fast, hold on a second, your module doesn't have same structure as mine, I don't want to have new tab added like this, check screenshot please

**Comment by khouloudbelguith** [comment]
@kpodemski, thanks for your feedback.
I tried with another module: https://github.com/PrestaShop/ps_qualityassurance using same structure.
I manage to reproduce the issue with PS1.7.6.7 and 177x branch.
![image](https://user-images.githubusercontent.com/16067358/92591278-e5f44680-f29d-11ea-9821-b38ba4c378dd.png)
I’ll add this to the debug roadmap so that it’s fixed. If you have already fixed it on your end or if you think you can do it, please do send us a pull request!
Thanks!

**Comment by jolelievre** [comment]
Hi @kpodemski 

I'd like to know how you build your tab structure? Do you rely on the `$tabs` property in your module?
I also have a couple questions, just to make sure I got your problem right:

`I don't want to have controller for parent tab`: I think this is done on purpose, from what I see in the BO menu all parents tabs could actually be independent Tabs (they all have related link and can be clicked), so that's why it is enforced and also why the validation prevents it
Your problem here is that you want a "non-clickable" parent so with no class nor route, if so then maybe instead of adding a `skip_validation` parameter we'd just need to allow empty Tabs that would be used for structure only (we would need o add a check on their depth maybe, to avoid having children Tabs with empty links)

Regarding your second question about the duplicateParent, here again (after some research I admit 😅 ) the method makes sense if we consider all tabs (including parents) need to be clickable. So when you add a child tab to a previously existing tab (no children), it actually creates a common parent tab and insert the new tab along with the existing one as children. What you would like here is to keep the existing tab as a parent without inserting its copy in children right? This implies that the parent link is not clickable among its children thus only the parent click allows to access this controller.

Sorry for the long summary, I just want to be sure I got it all right ^^ So basically we have a few features that need to be validated by the @PrestaShop/prestashop-product-team and we need to think of their hidden consequences ^^:
- is it authorized to have a parent Tab (first level of the menu) with no associated controller/route (so no link)
  - validation rules need to be changed and allow empty controller/route only for Tab at first level
  - If so I think during rendering the tab should be hidden if it doesn't have any children (in case they were removed or disabled)
  - when Tab are removed we need to check if the parent has no more children, if it also has no link then it should be removed
- is it possible to have a parent Tab (with a controller/route/link) that is not included in its own children (for now it's not)
  - if so then the new parameter `skip_duplication` seems like a good idea
  - we'll need to be extra careful with the `ModuleTabRegister::duplicateParentIfAlone` and `ModuleTabUnregister::removeDuplicatedParent`

**Comment by kpodemski** [comment]
@jolelievre you can take a look at my example module to see $tabs structure :)

**Comment by jolelievre** [comment]
Yep I just was it, I think the `skip_duplication` param would make more sense on the parent tab though (also we might find a better naming)

And I wouldn't recommend `skip_validation`, skipping validation is always the beginning of unexpected issues 😛  Updating the validation rules to match the new expected behaviour is a safer and more stable approach

**Comment by kpodemski** [comment]
@jolelievre 

yes, i think that we can just redirect in this parent controller to the one we want, thanks to "skip_duplication" we could create desired structure.

It's worth mentioning that registering tabs manually with `new Tab` works fine and won't create any duplicates, that's why i'm proposing this change.

> Yep I just was it, I think the skip_duplication param would make more sense on the parent tab though (also we might find a better naming)

We should have it in parent  and it's more intuitive but it's tricky due to the fact that at the beginning we don't know if there's a child or not for a given parent, and if not we should create this duplicate. 

For the naming, I was thinking about something like `as_parent`, or `as_main`, what do you think?

**Comment by jolelievre** [comment]
The way I saw it if we integrate this `skip_duplication` or `avoid_duplication` field it needs to be persisted in the Tab entity/table on creation So that you still have access to the information when you add a child later.

The drawback for this is that if you didn't previously set your parent Tab to avoid duplication then it's already too late, you can't define it when you add a child. But it kind of makes sense, it would be weird to leave to the child the decision of duplicating its parent or not The parent should be self configured and decide how it is supposed to behave, not its children (looks like real life use case when I read it 😅  )

Since this kind of architecture is likely to be within the scope of a single module it should work fine, the only case where it could be troublesome is in case the module is upgraded and the new structure is defined in the updated version. Then it might get tricky to update the existing Tab, but maybe it's acceptable that in these use cases the parent is duplicated.


I'm not sure I understood what you meant by `as_parent` `as_main` though? Is the idea to always have a link on parent tabs that would be filled with the "default" child (identified by `as_parent`)? I didn't think of this alternative, it would avoid having "empty" parent tabs indeed, but I guess using the first one would be acceptable and more dynamic.

NOTE: I just thought of a technical issue regarding the Tab that would have no controller/route I'm not sure how this will behave regarding permissions and user authorization since it's tightly connected to Tab, we will need to investigate but it might prevent us from creating "empty" Tabs.

**Comment by Prestaworks** [comment]
My take on this is that the "duplicateParentIfAlone" should be removed.
If I make a module and want the structure in the menu
A
-A
-B
I can just define that in the $tabs property, so I don't really see a reason to force the extra -A at all.

**Comment by boo-code** [comment]
The two things reported here are coupled in the code, which is worth stating before anyone picks a shape, because it decides whether one change is enough.

**Why a parent tab cannot be a container.** `ModuleTabRegister::checkIsValid()` requires every declared tab to resolve to something callable:

```php
$detectedControllers = $this->getDetectedModuleControllers($moduleName);
if (empty($data->get('route_name')) && !in_array($className, $detectedControllers)) {
    throw new Exception(sprintf('Class "%sController" not found in controllers/admin nor routing file', $className));
}
```

so a tab needs a `ModuleAdminController` or a `route_name`. A menu-only parent has neither.

**Why the duplicate then always appears.** `duplicateParentIfAlone()` runs on every `findParentId()`, and clones the parent as soon as its first child is added:

```php
if ($idParent === 0 || count($this->tabRepository->findByParentId($idParent))) {
    return $idParent;
}
```

Its purpose is sound: a parent that has its own link loses it once it gains a submenu, so the original is cloned into a container and pushed underneath. But because of the validation above, **every** parent has a link, so the clone always fires. @Prestaworks's `A / -A / -B` is exactly that: the extra `-A` is the original parent preserved.

So the two are one problem seen from both ends. Removing the duplication alone would silently break modules whose parent tab really is a page and would lose its link. Allowing a container parent alone would leave the duplication firing on it and produce the same extra entry.

The shapes proposed so far, with what each actually costs:

1. @jolelievre's `skip_duplication` persisted on the tab. Works for later children, and he already named its drawback: a parent registered before the flag existed does not have it.
2. @Prestaworks's "remove `duplicateParentIfAlone`". Simplest, and correct for container parents, but it is the only thing protecting a parent tab that is itself a page.
3. Derive it instead of declaring it: skip the duplication when the parent has no controller and no route, which is precisely the case where there is no link to preserve. That needs the validation relaxed for container tabs in the same change, and it needs no new field and no persistence, so an existing parent behaves correctly without having been marked.

I am not going to pick between these unilaterally, since it changes a module-facing contract in a way modules cannot detect. If one of them is the direction, say which and I will write it, validation and duplication together, since separating them does not produce a working result.
